emacs-everywhere.el

```diff
@@ -498,15 +549,39 @@ return windowTitle"))
 
 (defun emacs-everywhere-insert-selection ()
   "Insert the last text selection into the buffer."
-  (if (eq system-type 'darwin)
-      (progn
-        (call-process "osascript" nil nil nil
-                      "-e" "tell application \"System Events\" to keystroke \"c\" using command down")
-        (sleep-for 0.01) ; lets clipboard info propagate
-        (yank))
-    (when-let ((selection (gui-get-selection 'PRIMARY 'UTF8_STRING)))
-      (gui-backend-set-selection 'PRIMARY "")
-      (insert selection)))
+  (pcase system-type
+    ('darwin (progn
+	       (call-process "osascript" nil nil nil
+			     "-e" "tell application \"System Events\" to keystroke \"c\" using command down")
+               (sleep-for 0.01)	       ; lets clipboard info propagate
+               (yank)))
+    ((or 'ms-dos 'windows-nt 'cygwin)
+     (progn
+       (let* ((window-id (emacs-everywhere-app-id emacs-everywhere-current-app))
+	      (emacs-window-id (emacs-everywhere-call
+				"powershell"
+				"-NoProfile"
+				"-Command"
+				"& {Add-Type 'using System; using System.Runtime.InteropServices; public class Tricks { [DllImport(\"user32.dll\")] public static extern IntPtr GetForegroundWindow(); }'; [tricks]::GetForegroundWindow() }")))
+	 (apply #'call-process (car emacs-everywhere-window-focus-command)
+		nil nil nil
+		(mapcar (lambda (arg)
+                          (replace-regexp-in-string "%w" window-id arg))
+			(cdr emacs-everywhere-window-focus-command)))
+	 (apply #'call-process "powershell"
+                nil nil nil '("-NoProfile" "-Command" "& {(New-Object -ComObject wscript.shell).SendKeys('^c')}"))
+	 (sleep-for 0.1)
+
+	 (apply #'call-process (car emacs-everywhere-window-focus-command)
+		nil nil nil
+		(mapcar (lambda (arg)
+                          (replace-regexp-in-string "%w" emacs-window-id arg))
+			(cdr emacs-everywhere-window-focus-command))))
+       (sleep-for 0.1)
+       (yank)))
+    (_ (when-let ((selection (gui-get-selection 'PRIMARY 'UTF8_STRING)))
+	 (gui-backend-set-selection 'PRIMARY "")
+	 (insert selection))))
```

Comment:
For starters, `pcase` evaluation expression do not need to be wrapped with `progn`, you can do this just fine:
```elisp
(pcase expr
  (`condition
   (do-thing)
   (do-another-thing)
   ...))
```
Also, given how large the widows case is, I'd be tempted to create a new function called something like `emacs-everywhere-insert-selection--windows` to handle the windows-logic, and just call it here.

Reply:
Thanks for the tip! I'm exactly a starter.